### PR TITLE
feat: add local note graph

### PR DESCRIPTION
**Pull request by hfellerhoff**
This is a DRAFT PR 🚧. More testing will be done, as this is a fairly substantial addition to the V2 graph.

Changed/added:
- Added local note graph, a one-level-deep graph with direct hierarchical and linked connections from the current active note
- Made local note graph the default graph, with an option to view the full graph
- Enabled link connections on by default to fill out the local note graph
- Minor styling/rendering tweaks to enable graph switching and ease of use

### CODE REVIEW

File: packages/dendron-next-server/components/graph.tsx [hunk 1]
```diff
@@ -38,6 +40,12 @@ const getCytoscapeStyle = (themes: any, theme: string | undefined) => {
     background-color: ${AntThemes[theme].graph.node._selected.color};
     color: ${AntThemes[theme].graph.node._selected.color};
   }
+  
+  .parent {
+    shape: diamond;
+    width: ${AntThemes[theme].graph.node.size * 1.25};
```
Comment (kevinslin) [inline packages/dendron-next-server/components/graph.tsx:46]: how did we decide on the 1.25? should be a constant with a comment that describes reasoning

File: packages/dendron-next-server/components/graph.tsx [hunk 2]
```diff
@@ -112,6 +121,9 @@ export default function Graph({
 
   const { nodes, edges } = elements;
   const isLargeGraph = nodes.length + Object.values(edges).flat().length > 1000;
+  const isLocalGraph = config["options.show-local-graph"]
```
Comment (kevinslin) [inline packages/dendron-next-server/components/graph.tsx:124]: for easier re-use, lets make `is*Graph` logic into methods

File: packages/dendron-next-server/hooks/useGraphElements.tsx [hunk 3]
```diff
@@ -1,27 +1,267 @@
...
+        fnameArray[fnameArray.length - 2] ||
+        VaultUtils.getName(activeNote.vault);
+
+      const toVault = VaultUtils.getVaultByName({
+        vname: toVaultName,
+        vaults,
+      });
+
+      if (_.isUndefined(toVault)) {
+        logger.log(
```
Comment (kevinslin) [inline packages/dendron-next-server/hooks/useGraphElements.tsx:192]: i would classify this as an error log

File: packages/dendron-next-server/hooks/useGraphElements.tsx [hunk 4]
```diff
@@ -250,21 +495,50 @@ const useGraphElements = ({
   const [schemaCount, setSchemaCount] = useState(0);
 
   useEffect(() => {
+    const isLocalGraph = config["options.show-local-graph"]
```
Comment (kevinslin) [inline packages/dendron-next-server/hooks/useGraphElements.tsx:498]: see previous comment about making this a method

**Review by kevinslin — COMMENTED** [review]
nice progress here. let me know if this PR is still considered draft (it is according to the text in the PR but not in the status).

since we don't have automated tests yet, can you also add a checklist of manual test cases that you ran for this?

these test cases should be documented `pkg.dendron-next-server.qa` under `manual test cases`

File: packages/dendron-next-server/components/graph.tsx (hunk 1 above)
Comment (hfellerhoff) [reply packages/dendron-next-server/components/graph.tsx:46]: 1.25 is to compensate for the fact that diamond-shaped nodes appear smaller than circular ones. I'll refactor to be more specific

**Review by namjul — APPROVED** [review]

**Comment by namjul** [comment]
Only when I disable "Show Local Graph" the graph actually appears.
Enabling it again makes active note switching not being reflected in the graph.

https://www.loom.com/share/517d84cd2890498eae76657d54f77f95

File: packages/dendron-next-server/components/graph.tsx [hunk 5]
```diff
@@ -161,9 +186,18 @@ export default function Graph({
   };
 
   useEffect(() => {
+    // If changed from local graph to full graph, re-render graph to show all elements
+    const wasLocalGraph = cy && cy.$("node[localRoot]").length > 0;
```
[the remark this answers is not in the document]
Comment (hfellerhoff) [reply packages/dendron-next-server/components/graph.tsx:190]: > Only when I disable "Show Local Graph" the graph actually appears.
> Enabling it again makes active note switching not being reflected in the graph.
> 
> https://www.loom.com/share/517d84cd2890498eae76657d54f77f95

This was a simple fix for an issue I accidentally caused during some refactoring. It should be all good now

**Comment by hfellerhoff** [comment]
> Only when I disable "Show Local Graph" the graph actually appears.
> Enabling it again makes active note switching not being reflected in the graph.
> 
> https://www.loom.com/share/517d84cd2890498eae76657d54f77f95

This was a simple fix for an issue I accidentally caused during some refactoring. It should be all good now

**Comment by namjul** [comment]
> > Only when I disable "Show Local Graph" the graph actually appears.
> > Enabling it again makes active note switching not being reflected in the graph.
> > [loom.com/share/517d84cd2890498eae76657d54f77f95](https://www.loom.com/share/517d84cd2890498eae76657d54f77f95)
> 
> This was a simple fix for an issue I accidentally caused during some refactoring. It should be all good now

yes perfect works.

I found two other things
- opening the note-graph does not immediatly show the active note, you have to manually focus it again. (look into ShowPreviewV2 where this works and here where I fixed an issue with this https://github.com/dendronhq/dendron/pull/923)
- when switching "Show Local Graph" an error gets thrown

https://www.loom.com/share/db6926c43c3b41d5b8b4782eb60d52dd

**Comment by hfellerhoff** [comment]
> I found two other things
> 
> - opening the note-graph does not immediatly show the active note, you have to manually focus it again. (look into ShowPreviewV2 where this works and here where I fixed an issue with this https://github.com/dendronhq/dendron/pull/923)
> 
> - when switching "Show Local Graph" an error gets thrown
> 
> 
> 
> https://www.loom.com/share/db6926c43c3b41d5b8b4782eb60d52dd


I don't believe that the `next-server` package has access to the utils used in that ShowPreviewV2 fix - do you know of any that would be available in this package? I can take a look in a bit

**Comment by namjul** [comment]
> I don't believe that the next-server package has access to the utils used in that ShowPreviewV2 fix - do you know of any that would be available in this package? I can take a look in a bit

What I had in mind was that this section might be the one needed to be implemented.
https://github.com/dendronhq/dendron/blob/feat/local-graph/packages/plugin-core/src/commands/ShowNoteGraph.ts#L77-L93

**Comment by hfellerhoff** [comment]
> > I don't believe that the next-server package has access to the utils used in that ShowPreviewV2 fix - do you know of any that would be available in this package? I can take a look in a bit
> 
> What I had in mind was that this section might be the one needed to be implemented.
> https://github.com/dendronhq/dendron/blob/feat/local-graph/packages/plugin-core/src/commands/ShowNoteGraph.ts#L77-L93

Oh right, that makes complete sense - thanks for the heads up, I'll look into that

**Comment by hfellerhoff** [comment]
> I found two other things
> 
>     * opening the note-graph does not immediatly show the active note, you have to manually focus it again. (look into ShowPreviewV2 where this works and here where I fixed an issue with this [fix(preview): show current active note when exec ShowPreview command #923](https://github.com/dendronhq/dendron/pull/923))
> 
>     * when switching "Show Local Graph" an error gets thrown
> 
> 
> https://www.loom.com/share/db6926c43c3b41d5b8b4782eb60d52dd

Fixed issue 1, issue 2 is an error that gets thrown from within the Cytoscape library when the graph is re-rendered and hasn't finished rendering from a previous render. This error doesn't show in production, but I can look into it more if need be

File: packages/dendron-next-server/components/graph.tsx [hunk 6]
```diff
@@ -143,27 +165,55 @@ export default function Graph({
...
+        if (!isGraphLoaded) setIsGraphLoaded(true);
+      });
 
       network.on("select", (e) => onSelect(e));
 
       setCy(network);
     }
   };
 
+  // useEffect(() => {
```
Comment (kevinslin) [inline packages/dendron-next-server/components/graph.tsx:195]: do we still need this?

**Review by kevinslin — CHANGES_REQUESTED** [review]
great job. some minor issues:

- the information pane isn't updated with edges for local graph
- toggling from local graph to genreal graph loses the green highlight of the current note

**Review by kevinslin — CHANGES_REQUESTED** [review]
heads up that the integration tests are failing. can you fix that?

**Comment by hfellerhoff** [comment]
> heads up that the integration tests are failing. can you fix that?

Will do - I took a brief look yesterday and couldn't understand why `windowWatcher` was failing when there were no changes, but I'll take another stab at it

**Comment by kevinslin** [comment]
can you rebase from master? window watcher test can be flaky but currently
all tests are

**Comment by hfellerhoff** [comment]
> can you rebase from master? window watcher test can be flaky but currently all tests are
> […](#)

Should be good to go now

**Pull request merged**